### ISSUE TITLE
SAML login fails with PHP 7.3 ( Invalid expression at $dataObject = $xPath->query($query)->item(0))

### ISSUE THREAD

**Issue by jabouchleih** [nextcloud/user_saml#324]
### Steps to reproduce
1. Enable and configure the SSO & SAML application
2. Update to Nextcloud 15.0.7
3. Login via SAML

### Expected behaviour
The user gets redirected to the SSO page to enter his credentials.
After he entered his correct credentials, the SAML authentication should be successful and the user gets forwarded to the files page of Nextcloud.

### Actual behaviour
The user gets redirected to the SSO page to enter his credentials.
After he entered his correct credentials, an internal server error occurs:

```
The server was unable to complete your request.
If this happens again, please send the technical details below to the server administrator.
More details can be found in the server log.
Technical details
Remote Address: <SOME IP>
Request ID: <SOME ID>
```

### Server configuration
Latest docker image running on Debian 9.

**Operating system**:
Debian 9

**Web server:**
Reverse proxy (Apache 2)

**Database:**
MySQL

**PHP version:**
PHP 7.3.4 (cli) (built: Apr  6 2019 02:28:20) ( NTS )
Copyright (c) 1997-2018 The PHP Group
Zend Engine v3.3.4, Copyright (c) 1998-2018 Zend Technologies
    with Zend OPcache v7.3.4, Copyright (c) 1999-2018, by Zend Technologies

**Nextcloud version:** (see Nextcloud admin page)
15.0.7

**Where did you install Nextcloud from:**
Docker image (via docker-compose)

**List of activated apps:**
Enabled:
  - accessibility: 1.1.0
  - activity: 2.8.2
  - audioplayer: 2.6.1
  - calendar: 1.6.4
  - cloud_federation_api: 0.1.0
  - comments: 1.5.0
  - contacts: 3.1.1
  - dav: 1.8.1
  - drop_account: 0.1.0
  - federatedfilesharing: 1.5.0
  - federation: 1.5.0
  - files: 1.10.0
  - files_fulltextsearch: 1.2.6
  - files_markdown: 2.0.6
  - files_pdfviewer: 1.4.0
  - files_sharing: 1.7.0
  - files_texteditor: 2.7.0
  - files_trashbin: 1.5.0
  - files_versions: 1.8.0
  - files_videoplayer: 1.4.0
  - firstrunwizard: 2.4.0
  - fulltextsearch: 1.2.5
  - fulltextsearch_elasticsearch: 1.2.4
  - gallery: 18.2.0
  - jsloader: 1.0.2
  - logreader: 2.0.0
  - lookup_server_connector: 1.3.0
  - mail: 0.11.1
  - nextcloud_announcements: 1.4.0
  - notes: 2.6.0
  - notifications: 2.3.0
  - oauth2: 1.3.0
  - onlyoffice: 2.1.6
  - password_policy: 1.5.0
  - previewgenerator: 2.1.0
  - provisioning_api: 1.5.0
  - quota_warning: 1.4.0
  - serverinfo: 1.5.0
  - sharebymail: 1.5.0
  - socialsharing_email: 1.0.5
  - support: 1.0.0
  - survey_client: 1.3.0
  - systemtags: 1.5.0
  - tasks: 0.9.8
  - theming: 1.6.0
  - twofactor_backupcodes: 1.4.1
  - twofactor_totp: 2.1.2
  - updatenotification: 1.5.0
  - user_saml: 2.1.1
  - workflowengine: 1.5.0
Disabled:
  - admin_audit
  - bookmarks_fulltextsearch
  - circles
  - drawio
  - encryption
  - files_external
  - files_frommail
  - spreed
  - user_ldap

**Nextcloud configuration:**

```
{
    "system": {
        "trusted_proxies": "***REMOVED SENSITIVE VALUE***",
        "forwarded_for_headers": [
            "HTTP_X_REAL_IP"
        ],
        "auth.bruteforce.protection.enabled": false,
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "overwriteprotocol": "https",
        "trusted_domains": [
           <domains and ip list>
        ],
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "overwrite.cli.url": "<my domain>",
        "dbtype": "mysql",
        "version": "15.0.7.0",
        "dbname": "***REMOVED SENSITIVE VALUE***",
        "dbhost": "***REMOVED SENSITIVE VALUE***",
        "dbport": "",
        "dbtableprefix": "oc_",
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "installed": true,
        "memcache.local": "\\OC\\Memcache\\APCu",
        "filelocking.enabled": "true",
        "memcache.locking": "\\OC\\Memcache\\Redis",
        "redis": {
            "host": "***REMOVED SENSITIVE VALUE***",
            "port": 6379,
            "timeout": 0
        },
        "mail_smtpmode": "smtp",
        "mail_smtpauthtype": "LOGIN",
        "maintenance": false,
        "htaccess.RewriteBase": "\/",
        "log_type": "file",
        "logtimezone": "Europe\/Berlin",
        "loglevel": 2,
        "theme": "",
        "mail_smtpauth": 1,
        "mail_from_address": "***REMOVED SENSITIVE VALUE***",
        "mail_domain": "***REMOVED SENSITIVE VALUE***",
        "mail_smtphost": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpport": "465",
        "mail_smtpsecure": "ssl",
        "mail_smtpname": "***REMOVED SENSITIVE VALUE***",
        "mail_smtppassword": "***REMOVED SENSITIVE VALUE***",
        "updater.release.channel": "stable",
        "integrity.check.disabled": false,
        "onlyoffice": {
            "verify_peer_off": true
        },
        "apps_paths": [
            {
                "path": "\/var\/www\/html\/apps",
                "url": "\/apps",
                "writable": false
            },
            {
                "path": "\/var\/www\/html\/custom_apps",
                "url": "\/custom_apps",
                "writable": true
            }
        ]
    }
}

```
### Client configuration
**Browser:**
Google Chrome 73.0.3683.103
**Operating system:**
MacOS 10.14.4 (18E226) Mojave

### Logs

> [index] Error: Error: Call to a member function item() on bool at <<closure>>
> 0. 
 /var/www/html/custom_apps/user_saml/3rdparty/vendor/robrichards/xmlseclibs/src/XMLSecurityDSig.php line 591
>     processRefNode(DOMElement {})
>  1. /var/www/html/custom_apps/user_saml/3rdparty/vendor/onelogin/php-saml/src/Saml2/Utils.php line 1412
>     validateReference()
>  2. /var/www/html/custom_apps/user_saml/3rdparty/vendor/onelogin/php-saml/src/Saml2/Response.php line 410
>     validateSign(DOMDocument {}, "-----BEGIN CERT ... n", "", "sha1", "/samlp:Response/ds:Signature", null)
>  3. /var/www/html/custom_apps/user_saml/3rdparty/vendor/onelogin/php-saml/src/Saml2/Auth.php line 226
>     isValid("ONELOGIN_232389 ... 8")
>  4. /var/www/html/custom_apps/user_saml/lib/Controller/SAMLController.php line 251
>     processResponse("ONELOGIN_232389 ... 8")
>  5. /var/www/html/lib/private/AppFramework/Http/Dispatcher.php line 166
>     assertionConsumerService()
>  6. /var/www/html/lib/private/AppFramework/Http/Dispatcher.php line 99
>     executeController(OCA\User_SAML\Co ... {}, "assertionConsumerService")
>  7. /var/www/html/lib/private/AppFramework/App.php line 118
>     dispatch(OCA\User_SAML\Co ... {}, "assertionConsumerService")
>  8. /var/www/html/lib/private/AppFramework/Routing/RouteActionHandler.php line 47
>     main("OCA\\User_SAML\ ... r", "assertionConsumerService", OC\AppFramework\ ... {}, {_route: "user_s ... "})
>  9. <<closure>>
>     __invoke({_route: "user_s ... "})
> 10. /var/www/html/lib/private/Route/Router.php line 297
>     call_user_func(OC\AppFramework\ ... {}, {_route: "user_s ... "})
> 11. /var/www/html/lib/base.php line 987
>     match("/apps/user_saml/saml/acs")
> 12. /var/www/html/index.php line 42
>     handleRequest()
> 
> POST /apps/user_saml/saml/acs
> from <some ip> at 2019-04-17T08:34:13+02:00
> 
> [PHP] Error: DOMXPath::query(): Invalid expression at /var/www/html/custom_apps/user_saml/3rdparty/vendor/robrichards/xmlseclibs/src/XMLSecurityDSig.php#501
> 
> POST /apps/user_saml/saml/acs
> from <some ip> at 2019-04-17T08:34:13+02:00
> 
> [PHP] Error: preg_replace(): Compilation failed: invalid range in character class at offset 8 at /var/www/html/custom_apps/user_saml/3rdparty/vendor/robrichards/xmlseclibs/src/Utils/XPath.php#42
> 
> POST /apps/user_saml/saml/acs
> from <some ip> at 2019-04-17T08:34:13+02:00

 #### Browser log

> merged-share-backend.js?v=b20467e0-23:24 Uncaught ReferenceError: oc_appconfig is not defined
>     at merged-share-backend.js?v=b20467e0-23:24
>     at merged-share-backend.js?v=b20467e0-23:72
> (anonymous) @ merged-share-backend.js?v=b20467e0-23:24
> (anonymous) @ merged-share-backend.js?v=b20467e0-23:72
> 

> 12Cross-Origin Read Blocking (CORB) blocked cross-origin response <URL> with MIME type text/html. See <URL> for more details.

> merged-template-prepend.js?v=b20467e0-23:1527 Uncaught TypeError: Cannot read property 'substring' of undefined
>     at Object.filePath (merged-template-prepend.js?v=b20467e0-23:1527)
>     at viewer.js?v=b20467e0-23:25

The CORB messages seem a bit off, don't know if I have a misconfiguration somewhere or something changed between 15.x and 15.0.7.

The SAML provider is KeyCloak 5.

**Comment by ataraxus** [comment]
Same issue with my recent setup. 
Traefik + Keycloak (5.0.0) + NC 15.0.7:

`

Error | index | Error: Call to a member function item() on bool |   | 2019-04-19T23:11:23+0200
-- | -- | -- | -- | --
Error | PHP | DOMXPath::query(): Invalid expression at /var/www/html/custom_apps/user_saml/3rdparty/vendor/robrichards/xmlseclibs/src/XMLSecurityDSig.php#501 |   | 2019-04-19T23:11:23+0200
Error | PHP | DOMXPath::query(): Invalid expression at /var/www/html/custom_apps/user_saml/3rdparty/vendor/robrichards/xmlseclibs/src/XMLSecurityDSig.php#501 |   | 2019-04-19T23:11:23+0200
Error | PHP | preg_replace(): Compilation failed: invalid range in character class at offset 8 at /var/www/html/custom_apps/user_saml/3rdparty/vendor/robrichards/xmlseclibs/src/Utils/XPath.php#42


`

**Comment by Kosyne** [comment]
I'm getting the same issue as well. PHP 7.3, NC 15.07 (from zip) , Keycloak 6.0.0, on nginx, and behind an separate nginx reverse proxy.

Used this as reference when setting auth: https://stackoverflow.com/questions/48400812/sso-with-saml-keycloak-and-nextcloud

I believe my settings for `overwriteprotocol` and `overwritehost` are correct.

Can provide more info as needed. Keep searching and finding others with this issue, would love to get to the root of it.

**Comment by jabouchleih** [comment]
I also adjusted my `overwrite` settings, since that seemed to be an issue for some users but it still persists.

**Comment by ataraxus** [comment]
Hey Guys, I "fixed" it ... it seems to be related to the used PHP Version it breaks the SAML plugin

**Comment by jabouchleih** [comment]
I built my own docker image with PHP version 7.2. 
Everything works perfectly fine again - still this is a major issue. 

For everyone who doesn't know how to do that. Clone or download the official docker nextcloud repository at https://github.com/nextcloud/docker, navigatie to the 15.0/apache directory (or the one you need) and edit the `Dockerfile`. Replace the version at the beginning with 7.2.
```
# DO NOT EDIT: created by update.sh from Dockerfile-debian.template
FROM php:7.2-apache-stretch
```

**Comment by pfy** [comment]
user_saml 2.2.0 fixes the problem for us with php 7.3

**Comment by jabouchleih** [comment]
Fixed with v2.2.0 of the SSO & SAML plugin - closed.
